ouroboros-consensus/src/Ouroboros/Consensus/Node.hs

```diff
@@ -326,27 +328,23 @@ forkBlockProduction IS{..} =
           Just proof -> do
             (prevPoint, prevNo) <- prevPointAndBlockNo currentSlot <$>
                                      ChainDB.getCurrentChain chainDB
-
-            -- In this circumstance, it is required that we call 'syncState'
-            -- before 'getTxs' within this 'STM' transaction since we need to
-            -- guarantee that the transactions returned from 'getTxs' are
-            -- valid with respect to the current ledger state of the
-            -- 'ChainDB'. Refer to the 'getTxs' documentation for more
-            -- information.
-            _                   <- pure $ syncState mempool
-            mempoolSnapshot     <- getSnapshot mempool
-
-            let txs             =  map fst (snapshotTxs mempoolSnapshot)
             newBlock            <- runProtocol varDRG $
                                      produceBlock
                                        proof
                                        l
                                        currentSlot
-                                       (castPoint prevPoint)
+                                       prevPoint
                                        prevNo
-                                       txs
+                                       (map fst snapshotTxs)
             return $ Just newBlock
 
+      -- Note that there is a possible race condition here: we have produced a
+      -- block containing valid transactions w.r.t. the current ledger state
+      -- (this was race-free), but the current chain and consequently ledger
+      -- might change before we complete adding the block to the ChainDB. So
+      -- we might end up not adopting the block we produced ourselves because
+      -- it contained transactions that have become invalid since the time we
+      -- produced the block.
```

Comment:
It's not about txs becomming invalid, indeed they will not be. It's that if we add another block in the meantime then our new block will not make a longer chain than our new current chain, so it will not be adopted. But the new block we create will be perfectly valid, but chain selection will simply not select it. Lets fix this comment so we don't mislead ourselves later.

Reply:
Ah right, what I described is only secondary to this.